### PR TITLE
 [csrng/doc] application interface update

### PR DESCRIPTION
**Pull request by mwbranstad**
Signed-off-by: Mark Branstad <mark.branstad@wdc.com>

The application interface has been added to the markdown documentation.

### CODE REVIEW

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [inline hw/ip/csrng/doc/_index.md:154]: uninstantiate

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [inline hw/ip/csrng/doc/_index.md:149]: general comment - not sure if having 0x0 be a valid command value is ideal.  

I usually prefer to have them 1-hotted and/or value of "0" reserved / invalid.  Just my paranoia creeping in though :)

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [inline hw/ip/csrng/doc/_index.md:170]: sorry, is "(32b)" really meaning "(32b/word)" ?  Later you have "(128b)" and I think that is the total, so let's be consistent / more clear.

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [inline hw/ip/csrng/doc/_index.md:149]: Actually, is this bits 2-0 or bits 3-0.  What happened to bit 3?

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [inline hw/ip/csrng/doc/_index.md:168]: for all of these examples, I think I might suggest using Verilog syntax instead of C?  (4'h0, 4'h3, 4'hc, etc. for this example)

It's especially important if there are any non-nibble aligned subfields.

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [inline hw/ip/csrng/doc/_index.md:127]: strictly speaking are these csrng maintained states called "instances"?

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [inline hw/ip/csrng/doc/_index.md:278]: I think you need to define these 2 interrupts here.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:154]: @cdgori thanks for the spell check.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:149]: Considered this before. Zero will be invalid now, other commands increment by 1.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:170]: bits/word used in all places now.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:149]: acmd bit 3 implied as reserved before, now specified explicitly.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:168]: reasonable - all 0x0 notation replaced with verilog notation.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:127]: NIST refers to these states as a "working_state", and is referenced by a "state handle".  I have heard others refer to the state collection as instances, but there is no reference to state in this way by the NIST recommendation.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:278]: More of a placeholder, but the currently defined interests have now been added.

@cdgori thanks for the review!

File: hw/ip/csrng/doc/_index.md
Comment (imphil) [inline hw/ip/csrng/doc/_index.md:148]: This section duplicates some information from the section above in slightly different words (e.g. the description of xcnt). How about presenting this information in a table? E.g. something like that:

<table>
<thead>
  <tr>
    <th>bits</th>
    <th>name</th>
    <th>description</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>3:0</td>
    <td>acmd</td>
    <td>
      application command - selects one of five operations to perform

* 4'h0: invalid command
* 4'h1: instantiate command
* 4'h2: reseed command
* 4'h3: generate command
* 4'h4: update command
* 4'h5: uninstantiate command
* 4'h6-4'hF: reserved

    </td>
  </tr>
</table>

To make copy-past easier, here's the same as source code (you need to use HTML tables in markdown for half-sane multiline row support).


```

<table>
<thead>
  <tr>
    <th>bits</th>
    <th>name</th>
    <th>description</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>3:0</td>
    <td>acmd</td>
    <td>
      application command - selects one of five operations to perform

* 4'h0: invalid command
* 4'h1: instantiate command
* 4'h2: reseed command
* 4'h3: generate command
* 4'h4: update command
* 4'h5: uninstantiate command
* 4'h6-4'hF: reserved

    </td>
  </tr>
</table>
```

etc.

File: hw/ip/csrng/doc/_index.md
Comment (imphil) [inline hw/ip/csrng/doc/_index.md:171]: what's the "app number"? I cannot find a description in the doc.

File: hw/ip/csrng/doc/_index.md
Comment (imphil) [inline hw/ip/csrng/doc/_index.md:172]: Really "up to"?

File: hw/ip/csrng/doc/_index.md
Comment (imphil) [inline hw/ip/csrng/doc/_index.md:174]: Where are the command-specific flags defined?

File: hw/ip/csrng/doc/_index.md
Comment (imphil) [inline hw/ip/csrng/doc/_index.md:161]:
```suggestion
    - flgs: bits 15-12
```

File: hw/ip/csrng/doc/_index.md
Comment (imphil) [inline hw/ip/csrng/doc/_index.md:210]: Nit: Most of our existing ready/valid signals are actually using `_valid`, `_ready`, and `_data` as suffix.

**Comment by mwbranstad** [comment]
@imphil thanks for the review! I believe I have addressed all of your comments.

**Comment by mwbranstad** [comment]
@tjaychen tried to do a re-review but never had a first review completed on this PR.

File: hw/ip/csrng/doc/_index.md
Comment (tjaychen) [inline hw/ip/csrng/doc/_index.md:19]: should this snippet here mention the EDN?

File: hw/ip/csrng/doc/_index.md
Comment (tjaychen) [inline hw/ip/csrng/doc/_index.md:109]: just so i understand, this means the CSRNG can connect up to 15 EDNs, is that right?

File: hw/ip/csrng/doc/_index.md
Comment (tjaychen) [inline hw/ip/csrng/doc/_index.md:111]: this description is awesome.

File: hw/ip/csrng/doc/_index.md
Comment (tjaychen) [inline hw/ip/csrng/doc/_index.md:137]: i'm not quite sure why, but when i view this diagram, i see a few feint grey arrows pointing to nowhere.
Could they just be left over from a previous verion?

File: hw/ip/csrng/doc/_index.md
Comment (tjaychen) [inline hw/ip/csrng/doc/_index.md:148]: `coutn` -> `count`

File: hw/ip/csrng/doc/_index.md
Comment (tjaychen) [inline hw/ip/csrng/doc/_index.md:147]: wouldn't reseed potentially take some time also?

File: hw/ip/csrng/doc/_index.md
Comment (tjaychen) [inline hw/ip/csrng/doc/_index.md:152]: ah is this idea here if a block is making use of `block_encrypt` and not `update`, the `update` function can be taken by another command? And we may end up processing a couple of commands in parallel?

File: hw/ip/csrng/doc/_index.md
Comment (tjaychen) [inline hw/ip/csrng/doc/_index.md:202]: i think it would be good to mention this is the unmasked version, just for clarity.

**Comment by tjaychen** [comment]
this looks great Mark.  Thanks.
I've left a few clarifying comments, please have a look. 

I made some comments about mentioning EDN sooner.. but I now realize you probably did not intend for this block to always work with EDN.

**Review by tjaychen — APPROVED** [review]

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:19]: A bullet item in EDN was updated to mention the connections between peripherals, EDN, and CSRNG.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:109]: Correct.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:111]: and I meant to say earlier, thanks for this review @tjaychen !

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:137]: @rswarbrick has also reported graphics flaws. All of my views, both source and markdown inclusion, do not show any problems, so maybe @martin-lueker can take a look at this.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:148]: clearly a spell check bug...fixed.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:147]: True...but I was not included seed generation delay, which is true for both instantiate and reseed. This was referring to internal delays I supposed. The generate command could take multiple rounds of the AES ciper block versus a reseed command takes only one, for example.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:152]: Yes...sort of. The whole flow is basically a pipeline, so yes, while a generate is processing, there are holes where another command can sneak in. I guess indeed parallel in that case.

File: hw/ip/csrng/doc/_index.md
Comment (mwbranstad) [reply hw/ip/csrng/doc/_index.md:202]: Added.

File: hw/ip/csrng/doc/_index.md
Comment (tjaychen) [reply hw/ip/csrng/doc/_index.md:147]: sounds good.

File: hw/ip/csrng/doc/_index.md
Comment (tjaychen) [reply hw/ip/csrng/doc/_index.md:152]: sounds great!

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [inline hw/ip/csrng/doc/_index.md:123]: nit: ctr_drg_gen -> ctr_drbg_gen

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [inline hw/ip/csrng/doc/_index.md:206]: received -> receive

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [inline hw/ip/csrng/doc/_index.md:229]: additiona -> additional

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [inline hw/ip/csrng/doc/_index.md:323]: strictly speaking, it's the thing (hardware block or software) on the other end of the application interface that is responsible for reseeding, right?  said another way, the application interface logic doesn't do this autonomously, or at least I don't expect it to.

File: hw/ip/csrng/data/csrng.hjson
Comment (cdgori) [inline hw/ip/csrng/data/csrng.hjson:167]: FIPs->FIPS

**Review by cdgori — APPROVED** [review]
LGTM, modulo the few nits I commented on.  Thanks!

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:123]: Got it.  Thanks @cdgori.  Folding this in for @mwbranstad

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:137]: Yes.  For some reason the original Visio diagrams have these.  I can remove them.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:206]: Agree.  Got it Chris.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:229]: Got it.  Thanks again @cdgori

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:323]: Yes, this is the responsibility of the calling application.  For instance as described the EDN, docs.  (currently under review in #3391), that will be for us the most common attachment for this interface.  
Really what @mwbranstad has done here is created a full-featured interface to satisfy SP 800-90A, and then created the EDN to manage that according to our most common use-cases.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:212]: Hi @imphil,
How's this?
```
This section describes the application interface, which is required for performing any operations using a CSRNG 
instance (i.e. instantiation, reseeding, RNG generation, or uninstantiation).   Each CSRNG instance corresponds 
to a unique application interface port, which implements the application interface described here.  Any hardware 
peripherals which require complete control of an instance may connect directly to a dedicated interface port.
Meanwhile peripherals without any special requirements (i.e. personalization strings or non-FIPS-approved,
fully-deterministic number sequences) may share access to an instance via the entropy distribution network 
(EDN) IP. The EDN's manage the instantiation and reseeding of CSRNG instances for general use-cases, providing 
either on-demand or timed-delivery entropy streams to hardware peripherals. Firmware applications can obtain 
access to random bit sequences directly through application interface port 0, which is directly mapped to a set 
of TL-UL registers.
```

Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:212]: Done coming in next commit.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: Made a scan for unnecessary lines.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: Change made off line.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: Expanded a lot in the previous paragraph.  This sentence now reads:

"The total number of application interface ports (for TL-UL, directly attached peripherals or EDN instances) is determined by the `NHwApp` parameter."

(Application interface ports is defined in the previous paragraph.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: This appears to be out of date significantly.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: Table updated significantly in text.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: This table has been reformatted entirely since this comment was made.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: @mwbranstad's latest draft has done this.  (If he hasn't said this to you personally, he really expressed some significant approciation for this suggestion.)  It does look very nice with these new tables.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: Comment appears out of date with significant revisions.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: This section has significantly in the latest documentation tables.  Please let me know if this comment still applies.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: It seems this comment is out of date with the latest revision.

Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: Could you please look at the new tables?

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: It looks like @mwbranstad has taken care of this.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:19]: Thanks @mwbranstad. I've also added them here in response to @imphil's question about the intended users of the application interface.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:123]: Corrected. (By either myself or mwbranstad.)

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:206]: Corrected.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:229]: Corrected

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:323]: In the typical configuration this will be managed by the EDN.

File: hw/ip/csrng/data/csrng.hjson
Comment (martin-lueker) [reply hw/ip/csrng/data/csrng.hjson:167]: Good catch.  Thanks @cdgori.

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md]: Caught by @mwbranstad

File: hw/ip/csrng/doc/_index.md
Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:137]: Phantom Grey-lines removed, and this also has fixed the bounding-box problems.

File: hw/ip/csrng/doc/_index.md
Comment (cdgori) [reply hw/ip/csrng/doc/_index.md:323]: Sorry, I was not explicit here - the text says "it is the responsibility of the application interface to reseed" - I was suggesting it's really the calling application that is responsible, not the application interface itself.

Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:323]: Oh thanks for clarifying.  How does "...responsibility of the calling application..." sound?

Comment (cdgori) [reply hw/ip/csrng/doc/_index.md:323]: that's probably better and works for me.

Comment (martin-lueker) [reply hw/ip/csrng/doc/_index.md:323]: Also, I just realized that my previous comment was out of context... @mwbranstad is talking about tossing the `prediction_resistance_flag` out the window.  I'm not opposed to that I suppose, but we'll have to see what the lab says.

**Comment by martin-lueker** [comment]
Hello everyone.  Thank you so much for your review comments, they have made for a lot of improvements.   I've just pushed another commit: b6f3390224a9506e217a679b2bcd2393f9671626, which represents my attempt and catching any comments that have been missed in Mark's amends.

@imphil, many of the longer/paragraph changes (e.g. elaborating on the intended user of the application interface, description of the genbits interface, are in direct reponse to your broad comments).  Please take a look at this commit if you have a chance.

**Comment by martin-lueker** [comment]
Oh and @imphil, I just add one more line in d011248f1fc5ad2c58bb6d6d91258f395ac2fe94 about the application interface.   You said you wanted it in the "Overview section" I added it to the Overview/features section.  Is that what you had in mind?

**Review by rasmus-madsen — APPROVED** [review]

**Review by imphil — APPROVED** [review]

**Comment by imphil** [comment]
Thanks a lot Martin and Mark for writing up this document and the updates to it, it's a nice read now!

**Comment by imphil** [comment]
@martin-lueker Before merging, this the two commits in this PR should be combined with a commit message describing roughly what this whole PR does (doesn't need to be long, it's mostly human-readable text anyways). Can you do that, or should I do it and merge it then?

**Comment by martin-lueker** [comment]
> @martin-lueker Before merging, this the two commits in this PR should be combined with a commit message describing roughly what this whole PR does (doesn't need to be long, it's mostly human-readable text anyways). Can you do that, or should I do it and merge it then?

I can certainly do it.  Let me also check that the figure looks good on a linux browser. (@rswarbrick points out that there are font issues.)   I'll let everyone know when it is ready.

**Comment by martin-lueker** [comment]
Hi @imphil.  Thanks again for the positive feedback.  I am satisfied with the changes now.  Committers may merge when they are ready.

**Comment by imphil** [comment]
I updated this PR to change the commit author to Martin to match the signed-off-by line, and will merge as soon as CI passes.

**Comment by martin-lueker** [comment]
Thanks @imphil!   Just so I know for the future, what was this last problem?  (My previous commits didn't have CI problems, so my git authoring setup is getting things right at least sometimes.)

**Comment by imphil** [comment]
There was a mismatch between the commit author and the signed-off-by line. You combined all commits into one which was authored (according to git metadata) by Mark, but added your signed-off-by line. 

To fix that, I reset the commit author in the git metadata to match your signed-off-by line with `git commit --amend --author="Martin Lueker-Boden <martin.lueker-boden@wdc.com>"` (On your machine, a `git commit --amend -s --reset-author` would have worked as well.)

**Comment by martin-lueker** [comment]
I see.   It took Mark as the author because he made the first commit.   So I should have done reset-author.  Got it.  Thanks!

**Pull request merged**
